longtailstorelib/remotestore.go

```diff
@@ -384,49 +382,43 @@ func remoteWorker(
 				select {
 				case _ = <-flushMessages:
 					flushReplyMessages <- 0
-				case putMsg := <-putBlockMessages:
-					if accessType == ReadOnly {
-						putMsg.asyncCompleteAPI.OnComplete(longtaillib.EACCES)
-						continue
+				case putMsg, more := <-putBlockMessages:
+					if more {
+						if accessType == ReadOnly {
+							putMsg.asyncCompleteAPI.OnComplete(longtaillib.EACCES)
+							continue
+						}
+						err := putStoredBlock(ctx, s, client, blockIndexMessages, putMsg.storedBlock)
+						putMsg.asyncCompleteAPI.OnComplete(longtaillib.ErrorToErrno(err, longtaillib.EIO))
+					} else {
+						run = false
 					}
-					err := putStoredBlock(ctx, s, client, blockIndexMessages, putMsg.storedBlock)
-					putMsg.asyncCompleteAPI.OnComplete(longtaillib.ErrorToErrno(err, longtaillib.EIO))
 				case getMsg := <-getBlockMessages:
 					fetchBlock(ctx, s, client, getMsg)
 				case prefetchMsg := <-prefetchBlockChan:
 					prefetchBlock(ctx, s, client, prefetchMsg)
-				case _ = <-stopMessages:
-					run = false
 				}
 			} else {
 				select {
 				case _ = <-flushMessages:
 					flushReplyMessages <- 0
-				case putMsg := <-putBlockMessages:
-					if accessType == ReadOnly {
-						putMsg.asyncCompleteAPI.OnComplete(longtaillib.EACCES)
-						continue
+				case putMsg, more := <-putBlockMessages:
+					if more {
+						if accessType == ReadOnly {
+							putMsg.asyncCompleteAPI.OnComplete(longtaillib.EACCES)
+							continue
+						}
+						err := putStoredBlock(ctx, s, client, blockIndexMessages, putMsg.storedBlock)
+						putMsg.asyncCompleteAPI.OnComplete(longtaillib.ErrorToErrno(err, longtaillib.EIO))
+					} else {
+						run = false
 					}
-					err := putStoredBlock(ctx, s, client, blockIndexMessages, putMsg.storedBlock)
-					putMsg.asyncCompleteAPI.OnComplete(longtaillib.ErrorToErrno(err, longtaillib.EIO))
 				case getMsg := <-getBlockMessages:
 					fetchBlock(ctx, s, client, getMsg)
-				case _ = <-stopMessages:
-					run = false
 				}
 			}
 		}
 	}
-
-	for len(putBlockMessages) > 0 {
```

Comment:
This was wrong but not the actual cause of the bug - for putBlockMessages we always wait for a reply on the other end before asking the worker to stop so there couldn't be a case of dropped messages here.
Changed it to use the "more" flag since it is a sounder solution.